### PR TITLE
PE 433 Implement ANS 104 Bundles

### PR DESCRIPTION
**Pull request by jdaev**

### CODE REVIEW

File: lib/src/models/data-item.dart [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
 
-import 'package:json_annotation/json_annotation.dart';
+import 'package:arweave/src/utils/bundle_tag_parser.dart';
 
 import '../crypto/crypto.dart';
 import '../utils.dart';
 import 'models.dart';
 
-part 'data-item.g.dart';
+final MIN_BINARY_SIZE = 1044;
```
Comment (arielmelendez) [inline lib/src/models/data-item.dart:10]: Remove if unused

File: lib/src/models/data-item.dart [hunk 2]
```diff
@@ -105,46 +104,63 @@ class DataItem implements TransactionBase {
...
   @override
   Future<bool> verify() async {
+    final buffer = asBinary();
     try {
+      if (buffer.lengthInBytes < MIN_BINARY_SIZE) {
+        return false;
+      }
+      final sigType = byteArrayToLong(buffer.asUint8List().sublist(0, 2));
+      assert(sigType == 1);
+      var tagsStart = 2 + 512 + 512 + 2;
```
Comment (arielmelendez) [inline lib/src/models/data-item.dart:140]: Can we factor out all the magic numbers as constants to explain what they are based on the spec?

File: lib/src/utils.dart [hunk 3]
```diff
@@ -104,3 +105,84 @@ String getResponseError(Response res) {
 
 Future<String> ownerToAddress(String owner) async => encodeBytesToBase64(
     await sha256.hash(decodeBase64ToBytes(owner)).then((res) => res.bytes));
+
+Uint8List longTo32ByteArray(int long) {
```
Comment (arielmelendez) [inline lib/src/utils.dart:109]: Tech debt ticket for unit tests

Comment (arielmelendez) [reply lib/src/utils.dart:109]: Same for all below.

File: test/data_bundle_test.dart [hunk 4]
```diff
@@ -27,4 +29,33 @@ void main() async {
       expect(await dataItem.verify(), isFalse);
     });
   });
+
+  test('check if avro serializes tags correctly', () {
```
Comment (arielmelendez) [inline test/data_bundle_test.dart:33]: Whitespace between tests. Also consider testing the round trip -> raw string -> serialize -> deserialize -> check against raw string.

File: test/utils_test.dart [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import 'package:arweave/src/utils.dart';
+import 'package:test/test.dart';
+
+void main() async {
+  group('Utils:', () {
+    test('longTo8ByteArray', () async {
```
Comment (arielmelendez) [inline test/utils_test.dart:6]: Add some tests at the limits

**Review by arielmelendez — CHANGES_REQUESTED** [review]
Left a few comments

File: lib/src/models/data-item.dart (hunk 1 above)
Comment (jdaev) [reply lib/src/models/data-item.dart:10]: It's used. That's the min data item size.

File: test/utils_test.dart (hunk 5 above)
Comment (jdaev) [reply test/utils_test.dart:6]: [Created tech debt ticket ](https://ardrive.atlassian.net/jira/software/c/projects/PE/boards/2?modal=detail&selectedIssue=PE-595)

File: test/data_bundle_test.dart (hunk 4 above)
Comment (jdaev) [reply test/data_bundle_test.dart:33]: Done. 
[Also created tech debt ticket ](https://ardrive.atlassian.net/jira/software/c/projects/PE/boards/2?modal=detail&selectedIssue=PE-595)

File: lib/src/utils.dart (hunk 3 above)
Comment (jdaev) [reply lib/src/utils.dart:109]: [Created tech debt ticket ](https://ardrive.atlassian.net/jira/software/c/projects/PE/boards/2?modal=detail&selectedIssue=PE-595)

File: lib/src/models/data-item.dart (hunk 2 above)
Comment (jdaev) [reply lib/src/models/data-item.dart:142]: [Created Ticket](https://ardrive.atlassian.net/browse/PE-715)

File: lib/src/models/data-item.dart [hunk 6]
```diff
@@ -1,15 +1,14 @@
...
 
 import '../crypto/crypto.dart';
 import '../utils.dart';
 import 'models.dart';
 
-part 'data-item.g.dart';
+final MIN_BINARY_SIZE = 1044;
 
-@JsonSerializable()
 class DataItem implements TransactionBase {
```
[the remark this answers is not in the document]
Comment (jdaev) [reply lib/src/models/data-item.dart:14]: Done

**Review by arielmelendez — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -1,134 +1,144 @@
...
+
+To test, run the following command
+
+```shell
+dart test -p "chrome,vm"
+```
+
+## Development
+
+## Development
```
Comment (arielmelendez) [inline README.md:138]: Copy pasta?

File: test/data_bundle_test.dart [hunk 8]
```diff
@@ -23,7 +26,18 @@ void main() async {
       await dataItem.sign(wallet);
 
       expect(await dataItem.verify(), isTrue);
+    });
 
+    test('comfirm data item with wrong signaure fails verify', () async {
```
Comment (arielmelendez) [inline test/data_bundle_test.dart:31]: nit spelling: `comfirm` to `confirm`

Comment (arielmelendez) [inline test/data_bundle_test.dart:31]: What qualifies as a `wrong signature`? How does your test case result in a wrong signature?

File: test/data_bundle_test.dart [hunk 9]
```diff
@@ -34,7 +48,19 @@ void main() async {
     final buffer = serializeTags(tags: testTagsSnapshot);
     expect(buffer, equals(testTagsBufferSnapshot));
   });
-  
+
+  test('check if avro fails serialization when wrong data is given', () {
```
Comment (arielmelendez) [inline test/data_bundle_test.dart:52]: what qualifies as `wrong` other than the fact that the tags you provided literally have the words `wrong` involved?

File: test/data_bundle_test.dart [hunk 10]
```diff
@@ -60,4 +86,33 @@ void main() async {
...
+    await dataItemOne.sign(wallet);
+    final dataItemTwo =
+        DataItem.withBlobData(owner: await wallet.getOwner(), data: testData)
+          ..addTag('MyTag', '0')
+          ..addTag('OtherTag', 'Foo')
+          ..addTag('MyTag', '1');
+    await dataItemTwo.sign(wallet);
+    final bundle = DataBundle(items: [dataItemOne, dataItemTwo]);
+    await bundle.asBlob();
+    print(
```
Comment (arielmelendez) [inline test/data_bundle_test.dart:112]: Is this print still necessary?

File: README.md [hunk 11]
```diff
@@ -1,134 +1,142 @@
-# Arweave Dart SDK
```
Comment (arielmelendez) [inline README.md:1]: Why are there whitespace changes from top to bottom here?

**Review by arielmelendez — CHANGES_REQUESTED** [review]
Left a few questions and comments

File: test/data_bundle_test.dart (hunk 8 above)
Comment (jdaev) [reply test/data_bundle_test.dart:31]: Tags are a part of deriving the signature. What we are doing in this test is to create a data item and sign it. Then we append a new tag to the same data item. When we verify again the tags in the previous signature and the newly derived signature won't match. So it fails

File: test/data_bundle_test.dart (hunk 9 above)
Comment (jdaev) [reply test/data_bundle_test.dart:52]: We append a 0 to the buffer we pass in to the de serialize function. The tags themselves are irrelevant here.

File: README.md (hunk 7 above)
Comment (jdaev) [reply README.md]: Resolved

File: test/data_bundle_test.dart (hunk 9 above)
Comment (agsuy) [reply test/data_bundle_test.dart:52]: () => deserializeTags(buffer: [Uint8List.fromList(buffer), 0]) crashes the lib, there it is tested that lib crashes and we get an exception.

That being said, actual title should be something like "Test we properly catch Avro exceptions" or something like that.

File: README.md (hunk 11 above)
Comment (jdaev) [reply README.md:1]: VSCode markdown auto formatting. Should I discard these?

File: test/data_bundle_test.dart (hunk 10 above)
Comment (jdaev) [reply test/data_bundle_test.dart:112]: I think it's useful data. @agsuy

File: test/data_bundle_test.dart (hunk 8 above)
Comment (jdaev) [reply test/data_bundle_test.dart]: [Fixed](https://github.com/ardriveapp/arweave-dart/pull/10/commits/d9b05af369fd17d80febafe91ad7c79deea47ddd)

File: lib/src/models/tag.dart [hunk 12]
```diff
@@ -10,10 +10,11 @@ class Tag {
...
 
-  // TODO: Verify tag size.
-  // TODO: Move encoding logic into [Tag] class while ensuring it does not break JSON decoding.
   Tag(this.name, this.value);
 
   factory Tag.fromJson(Map<String, dynamic> json) => _$TagFromJson(json);
   Map<String, dynamic> toJson() => _$TagToJson(this);
+
+  @override
+  bool operator ==(o) => o is Tag && name == o.name && value == o.value;
```
Comment (matibat) [inline lib/src/models/tag.dart:19]: I wonder how the operator override looks like after it's compiled into JS :thinking:

File: lib/src/utils.dart [hunk 13]
```diff
@@ -104,3 +105,84 @@ String getResponseError(Response res) {
...
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+    0,
+  ];
```
Comment (matibat) [inline lib/src/utils.dart:144]: Is this possible here?
```suggestion
  final byteArray = new List(32);
```

File: lib/src/utils.dart [hunk 14]
```diff
@@ -104,3 +105,84 @@ String getResponseError(Response res) {
...
+  var value = 0;
+  for (var i = byteArray.length - 1; i >= 0; i--) {
+    value = value * 256 + byteArray[i];
+  }
+  return value;
+}
+
+Uint8List longTo8ByteArray(int long) {
+  // we want to represent the input as a 8-bytes array
+  final byteArray = [0, 0, 0, 0, 0, 0, 0, 0];
```
Comment (matibat) [inline lib/src/utils.dart:179]:
```suggestion
  final byteArray = new List(8);
```

File: lib/src/utils.dart [hunk 15]
```diff
@@ -104,3 +105,84 @@ String getResponseError(Response res) {
...
+    long = (long - byte) ~/ 256;
+  }
+
+  return Uint8List.fromList(byteArray);
+}
+
+Uint8List shortTo2ByteArray(int long) {
+  if (long > (2 ^ (32 - 1))) throw Exception('Short too long');
+
+  var byteArray = [0, 0];
```
Comment (matibat) [inline lib/src/utils.dart:158]:
```suggestion
  var byteArray = new List(2);
```

File: lib/src/utils/implementations/bundle_tag_parser_js.dart [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+        encodeBytesToBase64(
+            tag.name.split(',').map((e) => int.parse(e)).toList()),
+        encodeBytesToBase64(
+            tag.value.split(',').map((e) => int.parse(e)).toList()),
+      ));
+    }
+
+    return decodedTags;
+  } catch (e) {
+    throw WrongTagBufferException();
```
Comment (matibat) [inline lib/src/utils/implementations/bundle_tag_parser_js.dart:45]: Are we 100% sure that `e` is always because `WrongTagBufferException`?

File: lib/src/utils/implementations/bundle_tag_parser_js.dart [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+List<Tag> deserializeTags({var buffer}) {
+  try {
+    final tags = deserializeTagsFromBuffer(buffer);
+    final decodedTags = <Tag>[];
+    for (var tag in tags) {
+      decodedTags.add(Tag(
+        encodeBytesToBase64(
+            tag.name.split(',').map((e) => int.parse(e)).toList()),
```
Comment (matibat) [inline lib/src/utils/implementations/bundle_tag_parser_js.dart:37]: Isn't this already a `List`? Or maybe the `List` and `Array` are two different types?

File: test/fixtures/test_wallet.dart [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    "dp": "Gfo49fW5CZNTSEKQ_id0R2K9TMsoecw-jB2uCgqQi-TSLOtVRC5oTxA896my_SvIj8bCvEtLSzY3AhgvSCqulN3gSJbaHCCSDvAx0czAe7zfuTsxml76izeoKqg7TZAgAEnP0KXPRwJo4ff2J8lAcl3yyiLE7cLT9nuQSMRqERFVM7DQdk4wV618mQge9VGUStmYlh1MpS65N0dZWNafNuWauPTkTLZw8DFMIyizf3EC-nQYg1b6A_tYBHD3A82jPzQEQY8B3PrfGZ3DRASNv9jONk8qTQHOc5O5pLRMmUErDn_qRQCTKU483bzhooJE2a3WUEt6Pjsc1xMG4Vr3SQ",
+    "dq": "cCVai36Yi-06m1cwd8fbkhH9GUpXIvKI2Z5ZRk-smqc7piY0dEZFHftS9BaMyZYu3wM09GDklfdkNLo3mmfXkftv-cbjpvelUa50HYWx0HouKrT9UpVia0sTnmfme7BztjKunuuTcQxTBvfDfxoIi_nmUHIx9Vv1IEaALITzChGnIky3q7O_8ttKR65nFevG1JvsRBeJN6z0tzG9RBQr5mxtx3Wt2Uwcp21XjOCFHVmXjT9nMmpINQNNIC8VrGSSkjaJmNWIw5WGmDnLkKzCG2vpZO1suqIIgCsYN_Ka7ETTdZt3gFdoECUpFSiay4-4MAospvgWLv8XAFXXwfSPXQ",
+    "qi": "n-R81MpbwfWfqRSVgD8nDk7D8zlJ-tpMaojfTwNNqDt34Cr-BpMjxaQyEfMnzOd2dY4OV0rKhd29DIuwFEb2UERHdVWF3gM8f2byYGj4357CRkiwq6I050bUxd1ODgAXjVGNpOK_fmaNHDWfe5v3wVIcCmwH0mJxEu9kuz7fr9TJNxGJBGUphpGS6NQZDCbDXg9-FPafMeNV-Jdo0NQaKMwm8uZyW7YGSNpUXYnksrWt4Fa-B9H2KoC4PPSWESPxNooXdxK7Y0J1KbzNyrUmOl4dT6p_oFKcU-1unuDCZ11e6EmMKyUGjpDzTIAZ2XxmyWUJ06yzEw7oLo8noiCE_Q"
+  }''');
```
Comment (matibat) [inline test/fixtures/test_wallet.dart:17]: Is there a reason to have this JSON as a string then parse as a json? Maybe it is possible to have that as an object directly so we avoid parsing the exactly same json every time :thinking:

**Review by matibat — COMMENTED** [review]
Left some comments and questions there

File: test/fixtures/test_wallet.dart (hunk 18 above)
Comment (jdaev) [reply test/fixtures/test_wallet.dart:17]: Just a remainder from JY's implementation. The plan is to eventually move this to a github secret.

File: lib/src/utils/implementations/bundle_tag_parser_js.dart (hunk 17 above)
Comment (jdaev) [reply lib/src/utils/implementations/bundle_tag_parser_js.dart:37]: They are different types. In dart there's Iterables and Lists.

File: lib/src/utils/implementations/bundle_tag_parser_js.dart (hunk 16 above)
Comment (jdaev) [reply lib/src/utils/implementations/bundle_tag_parser_js.dart:45]: Yes, pretty much. Only reason that can fail is if the wrong data is passed in.

File: lib/src/utils.dart (hunk 14 above)
Comment (jdaev) [reply lib/src/utils.dart:179]: See answer above. I don't know if I should change this. I think seeing the list this way sort of gives you an immediate idea what's happening, but that could just be me.

File: lib/src/utils.dart (hunk 15 above)
Comment (jdaev) [reply lib/src/utils.dart:158]: See answer above.

**Review by fedellen — APPROVED** [review]
👌

**Review by agsuy — DISMISSED** [review]
Approved!

File: README.md (hunk 11 above)
Comment (jdaev) [reply README.md:1]: Tried to undo it, but it seems just adding that testing section made whitespace changes.

File: lib/src/utils.dart (hunk 13 above)
Comment (matibat) [reply lib/src/utils.dart:144]: From [here](https://api.dart.dev/stable/2.14.4/dart-core/List/List.filled.html), maybe:
```suggestion
  final byteArray = List.filled(32, 0, growable: false);
```

File: lib/src/utils.dart (hunk 14 above)
Comment (matibat) [reply lib/src/utils.dart:179]: It's not so hard for the dev to count 2 or 8 zeroes, but when it raises to 32 that's less trivial. See my comment above using `List.fill` as alternative to `List.generate`

File: README.md (hunk 11 above)
Comment (matibat) [reply README.md:1]: I'd say we better have the document formatted so it doesn't happen again in further changes

File: lib/src/utils.dart (hunk 13 above)
Comment (matibat) [reply lib/src/utils.dart:144]: We could actually use `List.generate` as we already are looping through the values to generate 'em in the loop below this line :thinking:

**Review by matibat — APPROVED** [review]
It mostly LGTM :+1:

**Review by arielmelendez — APPROVED** [review]
